stan/math/rev/fun/cholesky_decompose.hpp

```diff
@@ -358,76 +183,100 @@ class cholesky_opencl : public vari {
+  // looping gradient calcs faster for small matrices compared to
   // cholesky_block
   vari* dummy = new vari(0.0, false);
-  Eigen::Matrix<var, T::RowsAtCompileTime, T::ColsAtCompileTime> L(A.rows(),
-                                                                   A.cols());
+  arena_t<EigMat> L(L_A.rows(), L_A.cols());
   if (L_A.rows() <= 35) {
-    cholesky_scalar* baseVari = new cholesky_scalar(A_ref, L_A);
-    size_t accum = 0;
-    size_t accum_i = accum;
-    for (size_type j = 0; j < L.cols(); ++j) {
-      for (size_type i = j; i < L.cols(); ++i) {
-        accum_i += i;
-        size_t pos = j + accum_i;
-        L.coeffRef(i, j).vi_ = baseVari->vari_ref_L_[pos];
-      }
-      for (size_type k = 0; k < j; ++k) {
-        L.coeffRef(k, j).vi_ = dummy;
-      }
-      accum += j;
-      accum_i = accum;
-    }
+    internal::initialize_return(L, L_A, dummy);
+    reverse_pass_callback(internal::unblocked_cholesky_lambda(L_A, L, arena_A));
   } else {
 #ifdef STAN_OPENCL
     if (L_A.rows()
         > opencl_context.tuning_opts().cholesky_size_worth_transfer) {
-      cholesky_opencl* baseVari = new cholesky_opencl(A_ref, L_A);
-      internal::set_lower_tri_coeff_ref(L, baseVari->vari_ref_L_);
+      vari** vari_L = ChainableStack::instance_->memalloc_.alloc_array<vari*>(
+          arena_A.rows() * (arena_A.rows() + 1) / 2);
+      size_t pos = 0;
+      for (Eigen::Index j = 0; j < arena_A.rows(); ++j) {
+        for (Eigen::Index i = j; i < arena_A.rows(); ++i) {
+          vari_L[pos] = new vari(L_A.coeffRef(i, j), false);
+          L.coeffRef(i, j).vi_ = vari_L[pos];
+          ++pos;
+        }
+        for (Eigen::Index k = 0; k < j; ++k) {
+          L.coeffRef(k, j).vi_ = dummy;
+        }
+      }
+      reverse_pass_callback(internal::opencl_cholesky_lambda(arena_A, vari_L));
     } else {
-      cholesky_block* baseVari = new cholesky_block(A_ref, L_A);
-      internal::set_lower_tri_coeff_ref(L, baseVari->vari_ref_L_);
+      internal::initialize_return(L, L_A, dummy);
+      reverse_pass_callback(internal::cholesky_lambda(L_A, L, arena_A));
     }
 #else
-    cholesky_block* baseVari = new cholesky_block(A_ref, L_A);
-    internal::set_lower_tri_coeff_ref(L, baseVari->vari_ref_L_);
+    internal::initialize_return(L, L_A, dummy);
+    reverse_pass_callback(internal::cholesky_lambda(L_A, L, arena_A));
 #endif
   }
+  return plain_type_t<EigMat>(L);
+}
 
+/**
+ * Reverse mode specialization of Cholesky decomposition
+ *
+ * Internally calls Eigen::LLT rather than using
+ * stan::math::cholesky_decompose in order to use an inplace decomposition.
+ *
+ * Note chainable stack varis are created below in Matrix<var, -1, -1>
+ * @tparam T A `var_value` holding an inner eigen type.
+ * @param A A square positive definite matrix with no nan values.
+ * @return L Cholesky factor of A
+ */
+template <typename T, require_var_matrix_t<T>* = nullptr>
+inline auto cholesky_decompose(const T& A) {
```

Comment:
There should probably be an issue for `var<mat>` cholesky decompose not supporting OpenCL.

Reply:
It does exist: https://github.com/stan-dev/math/pull/2117#issue-comment-box

OpenCL version will no longer be ifdefed and will be completely separate in /opencl.

The ifdefed versions are going away once we support this in stanc3.

Reply:
Oh okay cool